src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs

```diff
@@ -314,25 +315,23 @@ private void ExecuteTestsInSource(IEnumerable<TestCase> tests, IRunContext? runC
                 }
 
                 var tasks = new List<Task>();
+                bool runSTA = MSTestSettings.RunConfigurationSettings.ExecutionApartmentState == ApartmentState.STA
+                    && RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
 
                 for (int i = 0; i < parallelWorkers; i++)
                 {
-                    tasks.Add(_taskFactory(() =>
+                    if (runSTA)
```

Comment:
I would not embed the logic here, in the ctor we have a factory we could check there if we're in STA and provide a factory working with custom STA thread and keep the code here transparent to the thread nature.